cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -362,50 +367,132 @@ __global__ void computeSplitClassificationKernel(
     col = select(colIndex, treeid, node.info.unique_id, seed, input.N);
   }
 
-  for (IdxT i = threadIdx.x; i < len; i += blockDim.x) shist[i] = 0;
+  // populating shared memory with initial values
+  for (IdxT i = threadIdx.x; i < pdf_shist_len; i += blockDim.x)
+    pdf_shist[i] = 0;
+  for (IdxT j = threadIdx.x; j < cdf_shist_len; j += blockDim.x)
+    cdf_shist[j] = 0;
   for (IdxT b = threadIdx.x; b < nbins; b += blockDim.x)
     sbins[b] = input.quantiles[col * nbins + b];
+
+  // synchronizing above changes across block
   __syncthreads();
+
+  // compute pdf shared histogram for all bins for all classes in shared mem
   auto coloffset = col * input.M;
-  // compute class histogram for all bins for all classes in shared mem
   for (auto i = range_start + tid; i < end; i += stride) {
+    // each thread works over a data point and strides to the next
     auto row = input.rowids[i];
     auto d = input.data[row + coloffset];
     auto label = input.labels[row];
+    // breakflag to check if data is greater than largest bin
+    bool breakflag = false;
     for (IdxT b = 0; b < nbins; ++b) {
-      auto isRight = d > sbins[b];  // no divergence
-      auto offset = b * 2 * nclasses + isRight * nclasses + label;
-      atomicAdd(shist + offset, 1);  // class hist
+      if (d <= sbins[b]) {  // shist (0 -> nbins*nclasses - 1)
```

Comment:
This is going to cause warp divergence, unlike the old code. What's the rationale for this change?

Reply:
IIUC, this seems to be a small price to pay for reducing smem atomic writes by an order of magitude...
So, previous code does `atomicAdd()` every single iteration (`O(n^2)`) . The change basically does it only once per outer data-sample and then breaks (`O(n)`).

Reply:
@hcho3 moreover, when we get #3606 I think this divergence will just be eliminated, even.